flink-streaming-java/src/test/java/org/apache/flink/streaming/util/SourceFunctionUtil.java

```diff
@@ -42,22 +41,32 @@
 public class SourceFunctionUtil {
 
 	public static <T extends Serializable> List<T> runSourceFunction(SourceFunction<T> sourceFunction) throws Exception {
-		final List<T> outputs = new ArrayList<T>();
-
 		if (sourceFunction instanceof RichFunction) {
+			return runRichSourceFunction(sourceFunction);
+		}
+		else {
+			return runNonRichSourceFunction(sourceFunction);
+		}
+	}
 
+	private static <T extends Serializable> List<T> runRichSourceFunction(SourceFunction<T> sourceFunction) throws Exception {
+		try (MockEnvironment environment = new MockEnvironment("MockTask", 3 * 1024 * 1024, new MockInputSplitProvider(), 1024)) {
 			AbstractStreamOperator<?> operator = mock(AbstractStreamOperator.class);
 			when(operator.getExecutionConfig()).thenReturn(new ExecutionConfig());
 
-			RuntimeContext runtimeContext =  new StreamingRuntimeContext(
-					operator,
-					new MockEnvironment("MockTask", 3 * 1024 * 1024, new MockInputSplitProvider(), 1024),
-					new HashMap<String, Accumulator<?, ?>>());
-
+			RuntimeContext runtimeContext = new StreamingRuntimeContext(
+				operator,
+				environment,
+				new HashMap<>());
 			((RichFunction) sourceFunction).setRuntimeContext(runtimeContext);
-
 			((RichFunction) sourceFunction).open(new Configuration());
+
+			return runNonRichSourceFunction(sourceFunction);
```

Comment:
The fact that `runRichSourceFunction` ended with a `runNonRichSourceFunction` was a bit surprising for me.

It maybe can be read better if we simply had this:
```
if (sourceFunction instanceof RichFunction) {
    setupRichSourceFunction(sourceFunction);
}
return runSourceFunction(sourceFunction);
```

where `setupUpSourceFunction` does all the operator / env / context mocking and invoking `setRuntimeContext`, `open` on the source function.

What do you think?

Reply:
It wouldn't suffice, because rich function's `MockEnvironment` must be closed. It would have to be something like:

```
Optional<MockEnvironment> context = Optional.empty();
try {
    if (sourceFunction instanceof RichFunction) {
        context = setupRichSourceFunction(sourceFunction);
    }
    return runSourceFunction(sourceFunction);
}
finally {
    context.ifPresent(MockEnvironment::close);
}
```

Which in my opinion is uglier :(

Reply:
Ah, I see. Thanks for the clarification, I didn't pay enough attention here.
Yes, it wouldn't make sense to refactor this any more if the environment needs to be closed.